Edgex-Source.go : Kuiperd panic on "close of closed channel" when a rule is dropped

**Environment**:

- eKuiper version (e.g. `1.3.4`): 
- Hardware configuration (e.g. `lscpu`):
- OS (mac):

**What happened and what you expected to happen**:
I have written a custom edgex source which is 90% same as the edge-source code provided in internal/topo. It runs completely fine and give the desired output but when I drop a rule after creating it, the kuiperd server panic and aborts. 
### The error is as follows:

panic: close of closed channel

goroutine 30 [running]:
github.com/lf-edge/ekuiper/internal/topo/connection/clients/edgex.(*edgexClientWrapper).unsubscribe(0xc0004b7080, 0x5004bf8, 0xc00073e0c0)
        github.com/lf-edge/ekuiper/internal/topo/connection/clients/edgex/edgex_wrapper.go:198 +0x32d
github.com/lf-edge/ekuiper/internal/topo/connection/clients/edgex.(*edgexClientWrapper).Release(0xc0004b7080, 0x5004bf8, 0xc00073e0c0)
        github.com/lf-edge/ekuiper/internal/topo/connection/clients/edgex/edgex_wrapper.go:217 +0x45
plugin/unnamed-2df0b11dd298b75df2589c0a09eeae8c907a7bc1.(*EdgexSource).Close(0xc0001c0190, 0x5004bf8, 0xc00073e0c0, 0xc000180360, 0xc000180420)
        command-line-arguments/customSource.go:308 +0x139
github.com/lf-edge/ekuiper/internal/topo/node.start.func1(0x5004bf8, 0xc0001e4540, 0x0, 0x1148c020, 0xc0001c0190, 0xc000158900)
        github.com/lf-edge/ekuiper/internal/topo/node/source_pool.go:307 +0xe7
created by github.com/lf-edge/ekuiper/internal/topo/node.start
        github.com/lf-edge/ekuiper/internal/topo/node/source_pool.go:303 +0x125

**How to reproduce it (as minimally and precisely as possible)**:
My custom edgex source code is:
`````
package main

import (
	"bytes"
	"encoding/gob"
	"encoding/json"
	"fmt"
	"strconv"
	"strings"

	"bitbucket.org/intechww/edge/pkg/common"
	"bitbucket.org/intechww/edge/pkg/models"
	v2 "github.com/edgexfoundry/go-mod-core-contracts/v2/common"
	"github.com/edgexfoundry/go-mod-core-contracts/v2/dtos"
	"github.com/edgexfoundry/go-mod-core-contracts/v2/dtos/requests"
	"github.com/edgexfoundry/go-mod-messaging/v2/pkg/types"
	"github.com/lf-edge/ekuiper/pkg/api"
	"github.com/lf-edge/ekuiper/pkg/cast"
)

type EdgexSource struct {
	cli api.MessageClient

	config            map[string]interface{}
	messageType       messageType
	buflen            int
	tags              map[string]string //key is tagFullName, value is alias
	receivedAllTags   bool
	receivedTagsCount int
	receivedTags      map[string]struct{} //key is tagFullName
}

type EdgexConf struct {
	Topic       string      `json:"topic"`
	MessageType messageType `json:"messageType"`
	BufferLen   int         `json:"bufferLength"`
}

type messageType string

const (
	MessageTypeEvent   messageType = "event"
	MessageTypeRequest messageType = "request"
)

func (es *EdgexSource) Configure(streamjson string, props map[string]interface{}) error {
	c := &EdgexConf{
		MessageType: MessageTypeEvent,
	}
	err := cast.MapToStruct(props, c)
	if err != nil {
		return fmt.Errorf("read properties %v fail with error: %v", props, err)
	}
	if c.BufferLen <= 0 {
		c.BufferLen = 1024
	}
	es.buflen = c.BufferLen
	es.messageType = c.MessageType
	es.config = props

	stream := models.Stream{}
	err = stream.UnmarshalJSON([]byte(streamjson))
	if err != nil {
		return err
	}
	es.tags = stream.Tags

	return nil
}

func (es *EdgexSource) Open(ctx api.StreamContext, consumer chan<- api.SourceTuple, errCh chan<- error) {
	log := ctx.GetLogger()

	cli, err := ctx.GetClient("edgex", es.config)
	if err != nil {
		errCh <- err
		log.Errorf("found error when get edgex client, error %s", err.Error())
		return
	}
	es.cli = cli.(api.MessageClient)

	messages := make(chan interface{}, es.buflen)
	es.receivedTags = make(map[string]struct{})
	es.receivedTagsCount = 0
	es.receivedAllTags = false

	topics := []api.TopicChannel{}
	for tag, _ := range es.tags {
		channelName, deviceName, tagName := common.StripTagName(tag)
		topic := "edgex/events/" + channelName + common.Separator + deviceName + "/" + channelName + common.Separator + deviceName + "/" + tagName
		topics = append(topics, api.TopicChannel{Topic: topic, Messages: messages})
	}

	subErrs := make(chan error, len(topics))
	if e := es.cli.Subscribe(ctx, topics, subErrs, nil); e != nil {
		log.Errorf("Failed to subscribe to edgex messagebus topic %s.\n", e)
		errCh <- e
	} else {
		log.Infof("Successfully subscribed to edgex messagebus topic %s.", topics)
		result := make(map[string]interface{})
		meta := make(map[string]interface{})
		for {
			select {
			case <-ctx.Done():
				log.Infof("Exit subscription to edgex messagebus topic %s.", topics)
				return
			case e1 := <-subErrs:
				errCh <- e1
				return
			case msg, ok := <-messages:
				if !ok { // the source is closed
					log.Infof("Exit subscription to edgex messagebus topic %s.", topics)
					return
				}

				env, ok := msg.(*types.MessageEnvelope)
				if !ok {
					log.Errorf("can not convert interface data to mqtt message.")
					return
				}

				if strings.EqualFold(env.ContentType, "application/json") {
					var r interface{}
					switch es.messageType {
					case MessageTypeEvent:
						r = &dtos.Event{}
					case MessageTypeRequest:
						r = &requests.AddEventRequest{}
					}
					if err := json.Unmarshal(env.Payload, r); err != nil {
						l := len(env.Payload)
						if l > 200 {
							l = 200
						}
						log.Warnf("payload %s unmarshal fail: %v", env.Payload[0:(l-1)], err)
					} else {
						// result := make(map[string]interface{})
						// meta := make(map[string]interface{})
						var e *dtos.Event
						switch t := r.(type) {
						case *dtos.Event:
							e = t
						case *requests.AddEventRequest:
							e = &t.Event
						}

						log.Debugf("receive message %s from device %s", env.Payload, e.DeviceName)
						for _, r := range e.Readings {
							tagAlias := ""
							if r.ResourceName != "" {
								if v, err := es.getValue(r, log); err != nil {
									log.Warnf("fail to get value for %s: %v", r.ResourceName, err)
								} else {
									tagAlias = es.tags[r.ResourceName]
									result[tagAlias] = v
									if !es.receivedAllTags {
										if _, ok := es.receivedTags[r.ResourceName]; !ok {
											es.receivedTags[r.ResourceName] = struct{}{}
											es.receivedTagsCount++
											fmt.Println(r.ResourceName)
											fmt.Println(es.receivedTagsCount)
											if es.receivedTagsCount == len(es.tags) {
												es.receivedAllTags = true
											}
										}
									}
								}
								r_meta := map[string]interface{}{}
								channel, device := common.StripDeviceName(r.DeviceName)
								r_meta["id"] = r.Id
								r_meta["origin"] = r.Origin
								r_meta["deviceName"] = device
								r_meta["channelName"] = channel
								r_meta["valueType"] = r.ValueType
								if r.MediaType != "" {
									r_meta["mediaType"] = r.MediaType
								}
								meta[tagAlias] = r_meta
							} else {
								log.Warnf("The name of readings should not be empty!")
							}
						}

						if len(result) > 0 && es.receivedAllTags {
							select {
							case consumer <- api.NewDefaultSourceTuple(deepCopyMap(result), deepCopyMap(meta)):
								log.Debugf("send data to device node")
							case <-ctx.Done():
								return
							}
						} else {
							log.Warnf("No readings are processed for the event, so ignore it.")
						}
					}
				} else {
					log.Errorf("Unsupported data type %s.", env.ContentType)
				}
			}
		}
	}
}

func (es *EdgexSource) getValue(r dtos.BaseReading, logger api.Logger) (interface{}, error) {
	t := r.ValueType
	logger.Debugf("name %s with type %s", r.ResourceName, r.ValueType)
	v := r.Value
	switch t {
	case v2.ValueTypeBool:
		if r, err := strconv.ParseBool(v); err != nil {
			return nil, err
		} else {
			return r, nil
		}
	case v2.ValueTypeInt8, v2.ValueTypeInt16, v2.ValueTypeInt32, v2.ValueTypeInt64, v2.ValueTypeUint8, v2.ValueTypeUint16, v2.ValueTypeUint32:
		if r, err := strconv.Atoi(v); err != nil {
			return nil, err
		} else {
			return r, nil
		}
	case v2.ValueTypeUint64:
		if u64, err := strconv.ParseUint(v, 10, 64); err != nil {
			return nil, err
		} else {
			return u64, nil
		}
	case v2.ValueTypeFloat32:
		if r, err := strconv.ParseFloat(v, 32); err != nil {
			return nil, err
		} else {
			return r, nil
		}
	case v2.ValueTypeFloat64:
		if r, err := strconv.ParseFloat(v, 64); err != nil {
			return nil, err
		} else {
			return r, nil
		}
	case v2.ValueTypeString:
		return v, nil
	case v2.ValueTypeBoolArray:
		var val []bool
		if e := json.Unmarshal([]byte(v), &val); e == nil {
			return val, nil
		} else {
			return nil, e
		}
	case v2.ValueTypeInt8Array, v2.ValueTypeInt16Array, v2.ValueTypeInt32Array, v2.ValueTypeInt64Array, v2.ValueTypeUint8Array, v2.ValueTypeUint16Array, v2.ValueTypeUint32Array:
		var val []int
		if e := json.Unmarshal([]byte(v), &val); e == nil {
			return val, nil
		} else {
			return nil, e
		}
	case v2.ValueTypeUint64Array:
		var val []uint64
		if e := json.Unmarshal([]byte(v), &val); e == nil {
			return val, nil
		} else {
			return nil, e
		}
	case v2.ValueTypeFloat32Array:
		return convertFloatArray(v, 32)
	case v2.ValueTypeFloat64Array:
		return convertFloatArray(v, 64)
	case v2.ValueTypeStringArray:
		var val []string
		if e := json.Unmarshal([]byte(v), &val); e == nil {
			return val, nil
		} else {
			return nil, e
		}
	case v2.ValueTypeBinary:
		return r.BinaryValue, nil
	case v2.ValueTypeObject:
		return r.ObjectValue, nil
	default:
		logger.Warnf("Not supported type %s, and processed as string value", t)
		return v, nil
	}
}

func convertFloatArray(v string, bitSize int) (interface{}, error) {
	var val1 []string
	if e := json.Unmarshal([]byte(v), &val1); e == nil {
		var ret []float64
		for _, v := range val1 {
			if fv, err := strconv.ParseFloat(v, bitSize); err != nil {
				return nil, err
			} else {
				ret = append(ret, fv)
			}
		}
		return ret, nil
	} else {
		var val []float64
		if e := json.Unmarshal([]byte(v), &val); e == nil {
			return val, nil
		} else {
			return nil, e
		}
	}
}

func (es *EdgexSource) Close(ctx api.StreamContext) error {
	log := ctx.GetLogger()
	log.Infof("EdgeX Source instance %d Done.", ctx.GetInstanceId())
	if es.cli != nil {
		es.cli.Release(ctx)
	}
	return nil
}

func init() {
	gob.Register(map[string]interface{}{})
}

// Map performs a deep copy of the given map m.
func deepCopyMap(m map[string]interface{}) map[string]interface{} {
	var buf bytes.Buffer
	enc := gob.NewEncoder(&buf)
	dec := gob.NewDecoder(&buf)
	enc.Encode(m)
	var copy map[string]interface{}
	dec.Decode(&copy)
	return copy
}

func CustomSource() api.Source {
	return &EdgexSource{}
}
`````
**Anything else we need to know?**:
Kindly help me in this regard.

--- a/.github/workflows/run_fvt_tests.yaml
+++ b/.github/workflows/run_fvt_tests.yaml
@@ -58,7 +58,7 @@ jobs:
     - name: install jmeter
       timeout-minutes: 10
       env:
-          JMETER_VERSION: 5.3
+          JMETER_VERSION: 5.4.3
       run: |
         wget --no-check-certificate -O /tmp/apache-jmeter.tgz https://downloads.apache.org//jmeter/binaries/apache-jmeter-$JMETER_VERSION.tgz
         cd /tmp && tar -xvf apache-jmeter.tgz
@@ -140,7 +140,7 @@ jobs:
       - name: install jmeter
         timeout-minutes: 10
         env:
-          JMETER_VERSION: 5.3
+          JMETER_VERSION: 5.4.3
         run: |
           wget --no-check-certificate -O /tmp/apache-jmeter.tgz https://downloads.apache.org//jmeter/binaries/apache-jmeter-$JMETER_VERSION.tgz
           cd /tmp && tar -xvf apache-jmeter.tgz
@@ -210,7 +210,7 @@ jobs:
     - name: install jmeter
       timeout-minutes: 10
       env:
-          JMETER_VERSION: 5.3
+          JMETER_VERSION: 5.4.3
       run: |
         wget --no-check-certificate -O /tmp/apache-jmeter.tgz https://downloads.apache.org//jmeter/binaries/apache-jmeter-$JMETER_VERSION.tgz
         cd /tmp && tar -xvf apache-jmeter.tgz

--- a/internal/topo/connection/clients/edgex/edgex_wrapper.go
+++ b/internal/topo/connection/clients/edgex/edgex_wrapper.go
@@ -106,6 +106,7 @@ func (mc *edgexClientWrapper) messageHandler(topic string, sub *edgexSubscriptio
 					for _, consumer := range sub.topicConsumers {
 						close(consumer.ConsumerChan)
 					}
+					conf.Log.Errorf("message handler for topic %s stopped", topic)
 					return
 				}
 				//broadcast to all topic subscribers
@@ -195,7 +196,6 @@ func (mc *edgexClientWrapper) unsubscribe(c api.StreamContext) {
 		if sub, found := mc.topicSubscriptions[tpc]; found {
 			for index, consumer := range sub.topicConsumers {
 				if strings.EqualFold(subId, consumer.ConsumerId) {
-					close(consumer.ConsumerChan)
 					sub.topicConsumers = append(sub.topicConsumers[:index], sub.topicConsumers[index+1:]...)
 					log.Infof("unsubscription topic %s for reqId %s, total subs %d", tpc, subId, len(sub.topicConsumers))
 				}

--- a/internal/topo/connection/clients/mqtt/mqtt_wrapper.go
+++ b/internal/topo/connection/clients/mqtt/mqtt_wrapper.go
@@ -203,7 +203,6 @@ func (mc *mqttClientWrapper) unsubscribe(c api.StreamContext) {
 		if sub, found := mc.topicSubscriptions[tpc]; found {
 			for index, consumer := range sub.topicConsumers {
 				if strings.EqualFold(subId, consumer.ConsumerId) {
-					close(consumer.ConsumerChan)
 					sub.topicConsumers = append(sub.topicConsumers[:index], sub.topicConsumers[index+1:]...)
 					log.Infof("unsubscription topic %s for reqId %s, total subs %d", tpc, subId, len(sub.topicConsumers))
 				}

--- a/internal/topo/transform/template.go
+++ b/internal/topo/transform/template.go
@@ -38,7 +38,7 @@ func GenTransform(dt string) (TransFunc, error) {
 			var output bytes.Buffer
 			err := tp.Execute(&output, d)
 			if err != nil {
-				return nil, false, fmt.Errorf("fail to encode data %v with dataTemplate", d)
+				return nil, false, fmt.Errorf("fail to encode data %v with dataTemplate for error %v", d, err)
 			}
 			return output.Bytes(), true, nil
 		} else {

